[stable10] FileTest::testPutWithModifyRun is testing a wrong behavior

DeepDiver1975 opened the pull request:
## Description
Due to the (as I assume) false annotation about the expected exception the real desired functionality is not tested (as show it is not working)

## How Has This Been Tested?
Run unit tests

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [x] Backport (if applicable set "backport-request" label and remove when the backport was done)

DeepDiver1975 committed:
FileTest::testPutWithModifyRun is testing a wrong behavior

DeepDiver1975 commented on apps/dav/tests/unit/Connector/Sabre/FileTest.php:
```diff
@@ -299,10 +299,8 @@ function ($path) use ($storage) {
 	 * from main function. The reason for exception is put from Sabre/File.php
 	 * throws exception
-	 *
-	 * @expectedException \Sabre\DAV\Exception
```
@sharidas mind having a look? What was the intention to expect this exception? THX

sharidas replied:
sure.

sharidas replied:
:man_facepalming: I will update the test. The idea behind the method was to modify the `run` in the *Hooks method. But this test is not doing its job. I will add my change along with this PR as a separate commit. Hope that sounds ok?

PVince81 commented:
expecting unit test fail then if it was expecting an exception, and got it, when now it isn't ?

DeepDiver1975 commented:
This Test throws an exception about fclose being called on a non resource.
All the asserts are never executed

DeepDiver1975 commented:
furthermore all there
```php
$file->put('test data');
```

don't work - the implementation of put expects a stream/resource to be passed in

there are multiple occations in this test - i wonder if there is more broken

sharidas committed:
Correct the faulty behaviour of testPutWithModifyRun

The following corrections were made to testPutWithModifyRun:
a) Its updated to listen to event file.beforeCreate eventually
   updated the assert
b) Updated the method with acquire and release lock so that
   the put method is executed gracefully.
c) Remove the listener of the event once we have tested in
   testPutWithModifyRun
d) The event name to be listened in testChunkedPutFails is
   adjusted accordingly

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas reviewed (commented):
I have added a commit which makes the testPutWithModifyRun behave correctly.

sharidas commented:
Test file when ran individually works:
```
➜  owncloud2 git:(bug-fix/wrong-test-behavior-FileTest-testPutWithModifyRun) ✗ make test-php TEST_DATABASE=pgsql TEST_PHP_SUITE=apps/dav/tests/unit/Connector/Sabre/FileTest.php 
PHPUNIT="/home/sujith/test/owncloud2/lib/composer/phpunit/phpunit/phpunit" build/autotest.sh pgsql apps/dav/tests/unit/Connector/Sabre/FileTest.php
Using PHP executable /usr/bin/php
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Parsing all files in lib/public for the presence of @since or @deprecated on each method...

Using database oc_autotest
Setup environment for pgsql testing on local storage ...
Installing ....
Cannot load Xdebug - it was already loaded
ownCloud was successfully installed
Testing with pgsql ...
Cannot load Xdebug - it was already loaded
/home/sujith/test/owncloud2/lib/composer/phpunit/phpunit/phpunit --configuration phpunit-autotest.xml --coverage-clover autotest-clover-pgsql.xml --coverage-html coverage-html-pgsql --log-junit autotest-results-pgsql.xml ../apps/dav/tests/unit/Connector/Sabre/FileTest.php 
Cannot load Xdebug - it was already loaded
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.7-1+ubuntu16.04.1+deb.sury.org+1 with Xdebug 2.6.0
Configuration: /home/sujith/test/owncloud2/tests/phpunit-autotest.xml

................................................................. 65 / 84 ( 77%)
...................                                               84 / 84 (100%)

Time: 39.42 seconds, Memory: 76.00MB

OK (84 tests, 472 assertions)

Generating code coverage report in Clover XML format ...^CMakefile:179: recipe for target 'test-php' failed
make: *** [test-php] Interrupt

➜  owncloud2 git:(bug-fix/wrong-test-behavior-FileTest-testPutWithModifyRun) ✗
```
Debugging further.

DeepDiver1975 commented:
nevertheless there are filing tests - please have a look

```
1) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFile
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:487

2) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "string" ('string', 0)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

3) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "castable string (int)" ('34', 34)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

4) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "castable string (float)" ('34.56', 34)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

5) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "float" (34.560000000000002, 34)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

6) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "zero" (0, 0)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

7) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "zero string" ('0', 0)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

8) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "negative zero string" ('-0', 0)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

9) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "string starting with number following by char" ('2345asdf', 2345)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

10) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "string castable hex int" ('0x45adf', 0)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

11) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "string that looks like invalid hex int" ('0x123g', 0)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

12) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "negative int" (-34, -34)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

13) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "negative float" (-34.43, -34)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

14) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "long int" (9223372036854775807, 9223372036854775807)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

15) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "too long int" (9.2233720368547758E+18, 9223372036854775807)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

16) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "long negative int" (-9223372036854775807, -9223372036854775807)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

17) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileLegalMtime with data set "too long negative int" (-9223372036854775808, -9223372036854775807)
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:573

18) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileTriggersHooks
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:644

19) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutSingleFileTriggersHooksDifferentRoot
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:470
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:739

20) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testUpdateFileWithPut
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:1210

21) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutLocking
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/tests/unit/Connector/Sabre/FileTest.php:1314

22) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\UploadTest::testBasicUpload
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/lib/Connector/Sabre/Directory.php:172
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:1095
/drone/src/lib/composer/sabre/dav/lib/DAV/CorePlugin.php:525
/drone/src/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:479
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:254
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/RequestTest.php:123
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/RequestTest.php:105
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/UploadTest.php:42

23) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\PartFileInRootUploadTest::testBasicUpload
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/lib/Connector/Sabre/Directory.php:172
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:1095
/drone/src/lib/composer/sabre/dav/lib/DAV/CorePlugin.php:525
/drone/src/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:479
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:254
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/RequestTest.php:123
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/RequestTest.php:105
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/UploadTest.php:42

24) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testBasicUpload
Sabre\DAV\Exception: Could not rename part file to final file

/drone/src/apps/dav/lib/Connector/Sabre/File.php:250
/drone/src/apps/dav/lib/Connector/Sabre/Directory.php:172
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:1095
/drone/src/lib/composer/sabre/dav/lib/DAV/CorePlugin.php:525
/drone/src/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:479
/drone/src/lib/composer/sabre/dav/lib/DAV/Server.php:254
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/RequestTest.php:123
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/RequestTest.php:105
/drone/src/apps/dav/tests/unit/Connector/Sabre/RequestTest/UploadTest.php:42

--

There was 1 failure:

1) OCA\Files_Versions\Tests\VersioningTest::testCopy
Failed asserting that false is true.

/drone/src/apps/files_versions/tests/VersioningTest.php:570

--

sharidas commented:
Now the only error appears to be in https://drone.owncloud.com/owncloud/core/8725/205 ( objectstore ).

phil-davis commented:
I have seen a lot of drone runs with 500 or more unit test fails in that job today. Something is wrong unrelated to this PR

sharidas commented:
@phil-davis The issue was with the symfony events not getting cleared/removed after the tests were completed. This was causing the issue. Now the errors are  with objectstore https://drone.owncloud.com/owncloud/core/8725/197. Rest of the errors which were troubling this PR are cleared.

DeepDiver1975 commented:
@ownclouders rebase

ownclouders commented:
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

sharidas commented on apps/dav/tests/unit/Connector/Sabre/FileTest.php:
```diff
@@ -296,14 +296,12 @@ function ($path) use ($storage) {
+		\OC::$server->getEventDispatcher()->addListener('file.beforeCreate', function (GenericEvent $event) use (&$calledUploadAllowed) {
+			$calledUploadAllowed[] = 'file.beforeCreate';
 			//Now modify run
 			$event->setArgument('run', false);
```
Well, the intention is not to terminate the operation. The intention is just to set the value to `false` if it is `true`. This test case was written for https://github.com/owncloud/core/blob/stable10/apps/dav/lib/Connector/Sabre/File.php#L311-L315. So basically we have a test to catch the event and set the value of run to false. Nothing else is expected from the test.

DeepDiver1975 replied:
well - if you look at the class View - there setting run to false causes the operation to terminate.

in https://github.com/owncloud/core/blob/stable10/apps/dav/lib/Connector/Sabre/File.php#L311-L315 the run value is not taken into account.

this is odd behavior - it smells ;-)

sharidas replied:
So in the updated PR, I have tried to unlock the file and then unlink it if the `$run` is false. Also would return null. Let me know if you are ok with the change?

PVince81 commented on apps/dav/lib/Connector/Sabre/File.php:
```diff
@@ -223,6 +223,10 @@ public function put($data) {
 			if ($view) {
 				$run = $this->emitPreHooks($exists);
+				if ($run === false) {
+					$view->unlockFile($this->path, ILockingProvider::LOCK_SHARED);
```
I don't see an acquireLock for this file above for some reason.

did we miss setting it in the first place ? :scream:

PVince81 replied:
also missing in 8.2: https://github.com/owncloud/core/blob/v8.2.8/lib/private/connector/sabre/file.php#L130

however the chunked code path is covered: https://github.com/owncloud/core/blob/v10.0.9/apps/dav/lib/Connector/Sabre/File.php#L502

PVince81 replied:
we could investigate this separately.

to move this PR forward, might need to remove this unlock as we might not have a lock applied. or please find the location where the shared lock is set to clarify

DeepDiver1975 replied:
There is the locking plugin which locks paths even before any other code is executed - https://github.com/owncloud/core/blob/eda2a55adfdecbe6af0f9e8ad589f878b64c1912/apps/dav/lib/Connector/Sabre/LockPlugin.php#L64

PVince81 replied:
now thinking, the plugin should probably also catch exceptions and release locks there as well... but in the end it's better to release as soon as possible.

I'm fine with this

PVince81 reviewed (approved):
:+1:

port to master ?

DeepDiver1975 commented:
> port to master ?

will take care ....
